### ISSUE TITLE
support slack notification 

### ISSUE THREAD

**Issue by alanEG** [opencve/opencve#85]
hello 
thanks you for this project 
can this tool support notification push to slack channel

**Comment by ncrocfer** [comment]
hello,
yes this is planned

**Comment by jhjacobs81** [comment]
Would this be by webhook? In which care Teams and others like MatterMost and RocketChat would be possible too. Not everyone uses slack or teams for privacy concerns or whatever :)

**Comment by ncrocfer** [comment]
Yes I think this "notification" feature must be generic, to allow multiple services to integrate with OpenCVE. Do you want to contribute to OpenCVE and do a PR ?

**Comment by ChriZzn** [comment]
Hi, 

any updates on this?
i plan to implement push notifications (webhook) in the next weeks....
@ncrocfer have you already thought about the implementation, or is there already preliminary work from you in this direction?

Regards Christoph

**Comment by ncrocfer** [comment]
Hi,

No concrete update to give you. I just began to "think about it" and try some tests in local. You can help us on this part, it's with very big pleasure !

Nevertheless here are some important details that have to be analysed before starting the dev.

## Pluggable Integrations

First of all some users want webhooks (OpenCVE executes a POST /endpoint request with some payload in it), but other users wants something different (raise an OpsGenie alert, create a JIRA ticket, send a message in a Slack channel, etc).

This list is not exhaustive, so we have to create an extensible feature, a kind of **pluggable integrations**, and not only focus on webhooks. Even an email will be an integration.

So the code has to handle this plugin system, and I imagine something like this:

```
opencve/
    api/
    commands/
    controllers/
    integrations/
       __init__.py
       email.py
       jira.py
       opsgenie.py
       slack.py
       webhook.py
       ...
     ...
```

The `__init__.py` file will contain the `BaseIntegration` class, and all integrations will inherit its methods (`get_client`, `send_request`, etc). This way it will be very simple to add new integrations and tests will be easier to write.

I think some integrations will be builtins (directly integrated in OpenCVE code) for *well-known* providers. But sometimes people would like to create their own integrations, too specific to be directly added in the core code. So OpenCVE has to handle the external integrations (something like `pip install opencve-foobar-integration` + an `[integrations]` new section in the `opencve.cfg` file).

## Security

Most integrations will have authentication (basic auth, simple token, JWT, etc). So OpenCVE **has** to encrypt this information in database.

I tried to use the [sqlalchemy_utils.StringEncryptedType](https://sqlalchemy-utils.readthedocs.io/en/latest/_modules/sqlalchemy_utils/types/encrypted/encrypted_type.html) class for that, with JSONType as underlying returned type, but it didn't worked well.

I created an issue for that 1 month ago: https://github.com/kvesteri/sqlalchemy-utils/issues/532 If you have an idea I'll take it :)

The model I created during my tests was something like that:

```python
# ... imports ...

class Integration(BaseModel):
    __tablename__ = "integrations"

    name = db.Column(db.String(), nullable=False)
    type = db.Column(db.String(), nullable=False)
    payload = db.Column(StringEncryptedType(JSONType(), AesEngine, "pkcs5"))

    # Relationships
    user_id = db.Column(UUIDType(binary=False), db.ForeignKey("users.id"))
    user = db.relationship("User", back_populates="integrations")

    def __repr__(self):
        return "<Integration {}>".format(self.name)
```

It was really a first test some days/weeks ago :) As you see the `type` column is a string. I first used an enum but it will be too complicated to create new migrations each time a new integration will be added, so a simple string will work.

## UI / UX

Each user will configure their integrations in the [Account > Notifications](https://www.opencve.io/account/notifications). I haven't thought about it yet, maybe a JSON config field (easier for UI, a textarea for instance). Or maybe dedicated fields per integration. I don't know yet. The second option is more interesting, but more complicated.

We have to define a limited number of integrations per user to limit the misuse of the tools (trust me, we have some users on the SaaS instance who created several thousands of subscriptions with fuzzing tools, certainly to *challenge* our platform. For sure it will be the same for the integrations, it's the game ^^).

This setting will be configured in the `opencve.cfg` file, so users with on-premise instance will be able to disable this limit.

## Tests

Finaly I think the integrations will be a central feature of OpenCVE, so of course everything has to be tests.

**Comment by ChriZzn** [comment]
Hi,

thanks for the Update,
You also thought about a "Global" Webhook Notification function...this would be mainly for on-prem installations, but would probably help most people?

**Comment by jhjacobs81** [comment]
Personally, i think most software now work with webhooks. I know JIRA does. We use it to hook our JIRA notifications into our Rocket.Chat instance :) i dont think it needs to be very complicated. But truthfully i’m no programmer, so what do i know ;-)

**Comment by ncrocfer** [comment]
@ChriZzn can you detail the architecture you have in mind please ? The answer I expect is for example: "as a user I want to be notified on OpsGenie, how can I create an integration on OpenCVE and configure it?"

I'm affraid your solution is only for a unique need, and later if someone else ask for a specific integration we'll not be able to answer him, so I will have to refactor your code and it will bring lot's of work.

@jhjacobs81 If I understand well your flow concerns JIRA -> Rocket.chat, here we talk to start from OpenCVE and go to a JIRA endpoint (OpenCVE -> JIRA). 

If I'm not wrong (but correct me please) OpenCVE webhooks solution will be to send a POST body to an endpoint, containing a payload, for example:

```
[{
  "cve": "CVE-2020-27223",
  "type": "references",
  "details": {
    "changed": [],
    "added": [{
      "url": "https://security.netapp.com/advisory/ntap-20210401-0005/",
      "name": "https://security.netapp.com/advisory/ntap-20210401-0005/",
      "refsource": "CONFIRM",
      "tags": []
    }],
    "removed": []
  }
}]
```

The keys are fixed by OpenCVE (`details`, `changed`, `references`, etc...), how can you configure JIRA to *intercept* this payload and create a new ticket ?

**Comment by varunnm17** [comment]
I was reading more about the usage of Slack through pluggable integrations rather than webhooks. I am curious about this approach. I wanted to ask if the plan you're describing still involves Slack bots or something else. If so, what is the difference between pluggable integrations and webhooks as I wanted to see which one would be easier to implement.

**Comment by zachw66** [comment]
Really excited about the integrations with Slack and possibly getting the JIRA integration working either through a standalone integration or embedded in the Slack integration (ex. Slack message is received on a subscribed product/vendor receiving new CVE or being updated with options to create a JIRA ticket for it). The integrations folder with a BaseIntegration class is a great idea.

I think it could also be a good idea to have a separate page under Accounts for managing all the integrations (Account > Integrations). I might be nice from a user perspective if every integration's settings was found in that one spot including settings for Slack notifications and then a small note could be made in Account > Notifications that Slack notification settings can be found in Integrations > Slack. For the UI, I'm thinking a similar UI style to the list of reports (https://www.opencve.io/reports) could be good. Obviously, there is a lot of ways that you could go with the UI. Anyways, those are just some thoughts; would love to hear some feedback and potentially help with working on these features.

**Comment by varunnm17** [comment]
Zach and I are working on a summer project and as a proof of concept for our business unit we are adding a slack bot.  Due to only having a few more weeks we won't be able to work on the fully integrated architecture you are proposing. 
We still wanted to share the settings we are going to add in case the knowledge is useful for the final product you are working on.

We wanted to include these settings on a per user bases:
* 		User Slack ID which will give information on who to send the CVE to
* 		CvSS Score (low, medium, high, critical) to see which type of CVE you would like to be informed upon
* 		Update Frequency where we would like to know if you would like to receive information every x minutes, daily, or weekly.
* 		Boolean value for alerting when older CVEs are updated.

We will share with you when we have a worked feature, we know it’s not the direction the project is going but it might still be useful for ideas.

**Comment by ncrocfer** [comment]
Hi @varunnm17 ,

Thank you for this proposal. To be fully honest I'm also working on this feature, based on the architecture I explained in a previous comment on this thread, and it will not just be a POC :)

I don't know what you will provide, and I hope I will be able to help you for your summer project, by reviewing your code for example, but I will be transparent: if you don't have time to finish it, and if in parallel I will already have a working solution, I will prefer to work on my own solution.

Just to explain me: it's easier to continue to work on a code I know rather than a POC providing by someone else. Plus you will provide a solution totally not integrated in the plugin architecture I have in mind, so later I will have to work to refactor your code to be compliant with the new plugin system, not an easy task and I don't want that.

But in any case if I can help you for your summer project I will do :)

**Comment by jhjacobs81** [comment]
Sorry for my late answer.. you dont let JIRA intercept anything.. you’d let OpenCVE do a POST om a JIRA webhook url. 

We do it the other way around:
Someone makes a ticket, JIRA posts a json thingy to a predefined webhook of RocketChat which shows up as a post from “JIRA BOT” in our alerts channel. 

But it can be the other way too. If something happens, you can let it automaticly create a ticket in JIRA by making an “incoming webhook” in JIRA settings. 

But instead of JIRA this could easily be Zabbix (monitoring solution) or Teams (slack alternative) or many other webapps, more and more apps are starting to support webhooks :)

**Comment by ChriZzn** [comment]
Hi there,

i had something like this in mind:
![image](https://user-images.githubusercontent.com/25209291/126195172-34e355d5-4c3f-479b-b6a7-b6ce1332a0d4.png)

And the format can be like you said: https://github.com/opencve/opencve/issues/85#issuecomment-875404868
this would be an "universal Solution" for every integration....in the worst case you need a little "data-broker" between the opencve webhook and lets say jira/slack/... or smth...but you can simply accomplish that.... (if you want we can provide examples for that)

this way would be the easiest and fastest and still everybody could do something with it...

How you think about that?

Regards Christoph

**Comment by ncrocfer** [comment]
Yes, this is exactly what I said in my previous and detailled answer: a "webhook" plugin.

**Comment by ChriZzn** [comment]
okay, when did you plan to implement the whole thing?
Before we all work on our own...wouldn't it be better to work together on this?

We would need this feature already, so I would like to ask my support, if necessary I could also give financial support?

Regards Christoph

**Comment by ncrocfer** [comment]
I have already started working on it, but because it's in my free time I don't have ETA to give you. Maybe next week, maybe in 3 weeks.

BTW thank you very much for the financial support, it's really appreciated, but here it's not the problem. The problem is: you insist to release a "global webhook" feature. While I would like something like this:

![opsgenie](https://user-images.githubusercontent.com/1552526/126330563-fda592c5-0451-428b-8e71-60f7862c7ba4.jpg)

A user-friendly interface where user can easily choose and configure their integrations. Do your feature will be able to provide that ? If yes I will really appreciate your contribution, for sure !

As you already said (or someone else, because this conversation started to be very long...) webhooks is now supported by a lot of solutions. If your feature can handle my previous request so go, please send me a PR. But this is an important feature for the project, so please provide a proper, tested and extensible solution (think of maintainability of the code).

An in any case **please** stop with your pressure on this topic !

**Comment by ChriZzn** [comment]
hi @ncrocfer 

another short update on this topic, we will now implement a global webhook function which can be globally enabled/disabled via the opencve.cfg and set a fixed webhook target (also via opencve.cfg), we will do a fork of this project for this, if you want you are welcome to merge this, but we also understand if you don't want this as it is not necessarily what you want

greetings christoph

**Comment by xme** [comment]
I was looking for a way to alert based on a webhook (let's try to reduce our addiction to emails! ;-) and I found this issue open for a while. What's the status of this idea? This could really improve the way automation could be implemented.

**Comment by henri9813** [comment]
Hello,

does this is still planned ?

Thanks !
